6412 Лаб.3 Ксеневич Максим

maeosha opened the pull request:
**ФИО:** Ксеневич Максим Станиславович
**Номер группы:** 6412
**Номер лабораторной:** 3  
**Номер варианта:** 10
**Краткое описание предметной области:** Поликилиника
**Краткое описание добавленных фич:**  Добелены EfCore репозитории, DbContex, Aspire проект. Данные хранятся в MySql с оркестрацией через Aspire.

alxmcs commented on StoreApp/StoreApp.Client/Program.cs:
Это что? Почему оно попало в репозиторий?

alxmcs commented on StoreApp/StoreApp.WebApplication/Pages/Customer.razor:
```diff
@@ -0,0 +1,191 @@
+﻿@page "/customers"
+@using StoreApp.StoreApp.Server.Dto
```
Я правильно понимаю, что тут происходит попытка зареференсить контракты из серверного приложения?
Удивительно, почему это не сработало
![image](https://github.com/user-attachments/assets/5160f2f3-76a2-4903-92b2-3e263d8d99ee)

alxmcs commented on StoreApp/StoreApp.WebApplication/appsettings.json:
Чтобы настройки из этого файла подтянусь в клиентское приложение, он должен находиться в совершенно другом месте

alxmcs reviewed (changes requested):
В итоге имеет следующее:
* сервер не работает
* клиент не билдится

alxmcs commented:
Итоги:
Все уже указано в ревью - сервер не работает, клиент не собирается. Ничего из этого кода зачесть нельзя. До встречи в феврале.

maeosha committed:
feat: add enums: BloodGroup, Gender and RhesusFactor

maeosha committed:
feat: add contract Appointment

maeosha committed:
feat: add reference book Specialisation

maeosha committed:
feat: add entities: Doctor and Patient

maeosha committed:
feat: add main class Clinic

maeosha committed:
feat: add dataseed for unit-tests

maeosha committed:
feat: add unit-tests

maeosha committed:
feat: add build configuration for C# project

maeosha committed:
feat: add solution

maeosha committed:
add summary and chage blood goup names

maeosha committed:
feat: extracted identical variables and functions from the doctors and patients classes into a separate class

maeosha committed:
feat: Now these classes inherit the basic characteristics from PersonInfo.
fix: add summary

maeosha committed:
fix: add summary

maeosha committed:
fix: Rewrite the tests and datasid, bring them to an adequate form.

maeosha committed:
fix: delete old Contracts/Appointment

maeosha committed:
fix: delete SpecialiSation.cs

maeosha committed:
feat: implement Patient management functionality with CRUD operations and DTOs

maeosha committed:
feat: add Specialization management functionality with CRUD operations and DTOs

maeosha committed:
feat: implement Doctor management functionality with CRUD operations and DTOs

maeosha committed:
feat: implement Appointment management functionality with CRUD operations and DTOs

maeosha committed:
feat: implement ClinicDataBase with CRUD operations for patients, doctors, specializations, and appointments

maeosha committed:
feat: add DataSeed class to populate the database with test data for patients, doctors, specializations, and appointments

maeosha committed:
feat: add DateConverter for JSON serialization of DateOnly type

maeosha committed:
feat: add AutoMapper profile for mapping between DTOs and entity models in the Clinic API

maeosha committed:
feat: initialize Clinic API with main program setup, including service registrations and Swagger integration

maeosha committed:
feat: add TestServices and TestControllers for enhanced patient and doctor data retrieval in the Clinic API

maeosha committed:
feat: update DataSeed

maeosha committed:
fix: change DateOnly to DateTime in GetAppointmentDto for correct date handling

maeosha committed:
fix: The Clinic.APi folder has been moved

maeosha committed:
fix: Now Specialization is also an entity

maeosha committed:
refactor: remove obsolete clinic model and associated test files

maeosha committed:
fix: update solution file to include Clinic.Api project and remove Clinic.Tests project

maeosha committed:
fix: remove custom mapping for DateTime in GetAppointmentDto

maeosha committed:
feat: add service interfaces for appointment, doctor, patient, and specialization management in the Clinic API

maeosha committed:
refactor: implement service interfaces for Appointment, Doctor, Patient, and Specialization services, enhancing method naming and documentation

maeosha committed:
feat: add base and specific controller interfaces for appointment, doctor, patient, and specialization management in the Clinic API

maeosha committed:
feat: add UpdateSpecialization method to ClinicDataBase and update interface to include specialization update functionality

maeosha committed:
refactor: implement base controller for appointment, doctor, patient, and specialization management, consolidating common CRUD operations and enhancing code reusability

maeosha committed:
refactor: update service registrations to use interfaces for Patient, Doctor, Specialization, and Appointment services, and add mapping for UpdateSpecializationDto

maeosha committed:
feat: implement Entity Framework data access classes for Appointment, Doctor, Patient, and Specialization management

maeosha committed:
feat: add interfaces for appointment, doctor, patient, and specialization data access with CRUD operations

maeosha committed:
Add initial database setup and seed data for clinic management system

- Created ClinicDbFactory for design-time DbContext instantiation.
- Added initial migration (20251219134315_InitialCreate) to set up database schema for Doctors, Patients, Specializations, Appointments, and DoctorSpecialization tables.
- Implemented seed data migration (20251219134336_SeedData) to populate initial records for specializations, doctors, patients, and appointments.
- Updated model snapshot to reflect the current state of the database schema.

maeosha committed:
feat: implement analytics controller and refactor existing controllers to use base controller pattern

maeosha committed:
Refactor clinic database structure and services

- Removed IClinicDataBase interface and its implementation.
- Deleted DataSeed class for test data.
- Updated Program.cs to configure Entity Framework and database context.
- Created AnalyticsServices for analytics-related functionalities.
- Refactored AppointmentServices, DoctorServices, PatientServices, and SpecializationServices to use specific data interfaces.
- Updated appsettings for database connection strings and logging levels.

maeosha committed:
feat: add unit tests and project configuration for Clinic.Tests

maeosha committed:
feat: add CI workflow for build and test automation

maeosha committed:
feat: configure Swagger UI and enable automatic launch in development environment

maeosha committed:
chore: remove outdated issue and discussion templates, and PR template

alxmcs commented on Clinic/Clinic.DataBase/EntityFramework/EfAppointmentDataBase.cs:
```diff
@@ -0,0 +1,106 @@
+/// Entity Framework implementation of <see cref="IAppointmentDataBase"/> that manages
+/// appointment entities using a <see cref="ClinicDbContext"/>.
+/// </summary>
+public sealed class EfAppointmentDataBase : IAppointmentDataBase
```
Здесь и далее по коду:
1. Название *DataBase явно не подходит ни классу, ни интерфейсу, который абстрагирует crud-операции с одной единственной коллекцией
2. Зачем помещать интерфейсы и реализации  в один и тот же инфраструктурный проект? Они должны использоваться как порты-адаптеры между разными слоями архитектуры

alxmcs commented on Clinic/Clinic.Api/DTOs/Appointment/CreateAppointmentDto.cs:
У меня сложилось впечатление, что половина моих замечаний с прошлого ревью было проигнорировано. К примеру, Create*Dto и Update*Dto все еще существуют и все еще полностью одинаковы

maeosha replied:
Ну они не полностью одинаковые, в UpdateDto поля не обязательные, можно обновить любое, а в CreateDto все поля required

alxmcs commented on Clinic/Clinic.Api/DTOs/Appointment/UpdateAppointment.cs:
Здесь и далее в аналогичных местах:
1. Название класса и название исходника не совпадают
2. Название неймспейса не соответствует структуре директорий, в которой класс находится

alxmcs reviewed (changes requested):
Самый главный вопрос - где оркестратор?
<img width="850" height="165" alt="image" src="https://github.com/user-attachments/assets/9ce30c8c-2f3e-454a-a339-9286e7c9c58b" />


Также со времен 2й лабы пропали инмемори репозитории. Если ты их ввел в свою архитектуру, будь добр поддерживать их до конца так же, как это делают все остальные студенты, которые сдавали 2 и 3 лабу отдельно.

maeosha commented:
Зачем поддерживать инмемори хранение, если смысл лабораторной работы как раз таки уйти от них?

alxmcs commented:
Философскими вопросами стоило задаваться в начале семестра, а не в 20х числах декабря.

Уход от использования инмемори репозиториев не означает отказ от их поддержки. Интерфейсы инфраструктурного уровня проекта нужно было спроектировать так, чтобы они оставались совместимы как с хранением в памяти, так и с хранением данных в бд через ef core.

maeosha commented:
Так про дальнейшую поддержку инмемори репозиториев не было сказано ни слова в условиях лабораторной работы...

maeosha committed:
feat: update DTO namespaces and add service defaults for improved structure and functionality

alxmcs commented on Clinic/Clinic.Tests/ClinicTests.cs:
<img width="1285" height="245" alt="image" src="https://github.com/user-attachments/assets/73044b90-50e1-4f84-986c-74dae8a1d7d6" />

alxmcs commented on Clinic/Clinic.Api/Clinic.Api.csproj:
```diff
@@ -0,0 +1,30 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>net9.0</TargetFramework>
```
Я уже устал каждое ревью писать, что фреймворк должен быть 8 версии

alxmcs commented on Clinic/Clinic.AppHost/Clinic.AppHost.csproj:
```diff
@@ -0,0 +1,19 @@
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net9.0</TargetFramework>
```
Аналогично

alxmcs commented on Clinic/Clinic.DataBase/Clinic.DataBase.csproj:
```diff
@@ -0,0 +1,23 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net9.0</TargetFramework>
```
Аналогично

alxmcs commented on Clinic/Clinic.Models/Clinic.Models.csproj:
```diff
@@ -0,0 +1,9 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net9.0</TargetFramework>
```
Аналогично

alxmcs commented on Clinic/Clinic.ServiceDefaults/Clinic.ServiceDefaults.csproj:
```diff
@@ -0,0 +1,22 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net9.0</TargetFramework>
```
Аналогично

alxmcs commented on Clinic/Clinic.Tests/Clinic.Tests.csproj:
```diff
@@ -0,0 +1,32 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net9.0</TargetFramework>
```
Аналогично

alxmcs reviewed (changes requested):
Как-то поздно я, конечно, заинтересовался вопросом. Но почему вариант 10?
<img width="368" height="158" alt="image" src="https://github.com/user-attachments/assets/719dbf19-d1ac-415b-b507-50efd84c75ef" />

Когда должен быть 36
<img width="449" height="310" alt="image" src="https://github.com/user-attachments/assets/3c2cc6d3-fd0f-4b6a-8a3c-6b9af5b033ae" />

Предыдущие лабы, я, к сожалению, уже принял, так что заставлять тебя менять доменную область было бы нечестно. Но данную лабу я принимать не буду, пока не будет сменено СУБД, потому что по варианту должен быть Postgres, а не MySql.
<img width="652" height="47" alt="image" src="https://github.com/user-attachments/assets/2d94eb80-3191-4d17-b51b-d661b39ffb78" />

Кстати, [об отсутствии в условии лабораторных необходимости поддерживать инмемори репозитории](https://github.com/itsecd/enterprise-development/discussions/87#discussioncomment-14837255).

maeosha committed:
feat: refactor repository interfaces and implementations for better structure and add Clinic.Application project

maeosha committed:
fix: changed the database from MySQL to PostgreSQL

maeosha commented:
Пожалуйста сжальтесь, дайте возможность не возвращать мне инмемори репозитории, мне под них инфрастурктуру менять придется:(

alxmcs commented:
Лол, нет

maeosha committed:
feat: add in-memory repositories for appointments, doctors, patients, and specializations

maeosha committed:
Fix project SDK declaration in csproj file

alxmcs commented on Clinic/Clinic.Api/Controllers/SpecializationControllers.cs:
```diff
@@ -0,0 +1,13 @@
+/// </summary>
+[ApiController]
+[Route("api/specializations")]
+public class SpecializationControllers(ISpecializationServices specializationServices) : BaseControllers<GetSpecializationDto, CreateSpecializationDto, UpdateSpecializationDto>(specializationServices);
```
Документации ни к методам контроллера
<img width="399" height="521" alt="image" src="https://github.com/user-attachments/assets/54c5f6c9-e4e4-422e-8170-0aa618b0cac1" />

alxmcs commented on Clinic/Clinic.Api/DTOs/Appointment/CreateAppointmentDto.cs:
Ни к контрактам нет

alxmcs commented on Clinic/Clinic.Api/Clinic.Api.csproj:
```diff
@@ -0,0 +1,21 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
```
Потому что ее никто не билдит

alxmcs commented on Clinic/Clinic.Api/Program.cs:
```diff
@@ -0,0 +1,64 @@
+builder.Services.AddAutoMapper(cfg => cfg.AddProfile<MappingProfile>());
+
+builder.Services.AddEndpointsApiExplorer();
+builder.Services.AddSwaggerGen();
```
И не добавляет в сваггерген

alxmcs commented on Clinic/Clinic.Models/Entities/Specialization.cs:
```diff
@@ -0,0 +1,19 @@
+/// <summary>
+/// Represents a medical specialisation in the clinic.
+/// </summary>
+public class Specialization
+{
+    /// <summary>
+    /// Gets or sets the id specialization.
+    /// </summary>
+    required public int Id { get; set; }
```
Ни при помощи атрибутов

alxmcs commented on Clinic/Clinic.DataBase/ClinicDbContext.cs:
```diff
@@ -0,0 +1,108 @@
+        // Configure Patient entity
+        modelBuilder.Entity<Patient>(entity =>
+        {
+            entity.HasKey(e => e.Id);   
+            entity.Property(e => e.PassportNumber).IsRequired().HasMaxLength(50);
```
Ни при помощи fulent api именования таблиц и столбцов не приводятся к конвенциям именования postgres

alxmcs commented on Clinic/Clinic.Api/Controllers/AnalyticsControllers.cs:
```diff
@@ -0,0 +1,76 @@
+/// </summary>
+[ApiController]
+[Route("api/analytics")]
+public class AnalyticsController(AnalyticsServices testServices) : ControllerBase
```
Контроллеры в проекте почему-то все проименованы во множественном числе

alxmcs commented on Clinic/Clinic.Api/DTOs/Appointment/CreateAppointmentDto.cs:
```diff
@@ -0,0 +1,33 @@
+/// </summary>
+public class CreateAppointmentDto
+{
+    /// <summary>
+    /// The full name of the patient for the appointment.
+    /// </summary>
+    public required string PatientFullName { get; set; }
+
+    /// <summary>
+    /// The full name of the doctor for the appointment.
+    /// </summary>
+    public required string DoctorFullName { get; set; }
```
Почему для связи с другими сущностями тут дважды используется неуникальное поле, которое приведет к конфликтам и рассогласованности данных, я понятия не имею

alxmcs commented on Clinic/Clinic.Api/DTOs/Appointment/UpdateAppointment.cs:
```diff
@@ -0,0 +1,33 @@
+    /// <summary>
+    /// Optional: The ID of the patient to update for the appointment.
+    /// </summary>
+    public int? PatientId { get; set; } = null;
+
+    /// <summary>
+    /// Optional: The ID of the doctor to update for the appointment.
+    /// </summary>
+    public int? DoctorId { get; set; } = null;
```
Почему-то здесь здравый смысл все-таки одержал верх, а в `CreateAppointmentDto` - нет

alxmcs commented on Clinic/Clinic.Api/Interfaces/Services/IAppointmentService.cs:
```diff
@@ -0,0 +1,25 @@
+using Clinic.Api.DTOs.Appointment;
+
+namespace Clinic.Api.Interfaces.Services;
```
Весь семестр мы говорили о грамотной декомпозиции кода проекта, о разделении слоев, о логической связности, о domain-centered grouping

Не, зачем, тут будет папка "интерфейсы" и в ней будут валяться все интерфейсы 👍 👍 👍 👍

alxmcs commented on Clinic/Clinic.Application/Ports/IAppointmentRepository.cs:
```diff
@@ -0,0 +1,52 @@
+using Clinic.Models.Entities;
+
+namespace Clinic.Application.Ports;
```
Архитектурная концепция портов-адаптеров никогда не подразумевала под собой помойку из интерфейсов в отдельной директории с названием "порты"

alxmcs reviewed (approved):
Проект обрастает функционалом, а лучше от этого никому не становится. В том числе и мне. К 4 лабораторной у меня, наверное, окончательно сосуды в мозгу лопнут.

Я уже несколько ревью подряд пишу о необходимости выделить dto в отдельную класслибу. Ты как планируешь реализовывать совершенно автономный и никак не связанный с апи сервером сервер генерации контрактов без отдельной либы с контрактами? Апи сервер в нем зареференсишь? А если бы нужно было писать фронтенд, ты бы в клиентском приложении сервер зареференсил?

Может быть, вместо бокса по переписке под конец семестра стоило начать писать код пораньше (и получше)?
